### PR TITLE
Add optimizations for ZBB extension

### PR DESCRIPTION
**Pull request by djtodoro**

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -7579,6 +7579,27 @@ static SDValue performFP_TO_INT_SATCombine(SDNode *N,
   return DAG.getSelectCC(DL, Src, Src, ZeroInt, FpToInt, ISD::CondCode::SETUO);
 }
 
+static SDValue foldSelectofCTTZ(SDNode *N, SelectionDAG &DAG) {
```
Comment (draganmladjenovic) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:7582]: Is this a bit lax? What exact pattern do we wish to fold here

Comment (djtodoro) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:7582]: Well. It needs more checks for sure. Thanks.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 2]
```diff
@@ -7579,6 +7579,48 @@ static SDValue performFP_TO_INT_SATCombine(SDNode *N,
...
+  SDValue CTTZ = N->getOperand(4);
+  if (!CTTZ.getNumOperands())
+    return SDValue();
+
+  if (CTTZ.getOpcode() != ISD::CTTZ && CTTZ.getOpcode() != RISCVISD::CTZW)
+    return SDValue();
+
+  SDValue FalseV = N->getOperand(4);
+  // The `false` branch should be CTTZ itself.
+  if (FalseV != CTTZ)
```
Comment (draganmladjenovic) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: Always false given the line 7586.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 3]
```diff
@@ -7579,6 +7579,48 @@ static SDValue performFP_TO_INT_SATCombine(SDNode *N,
...
+    return SDValue();
+
+  //  The `true` branch should be constant `0`, `31` or `63`.
```
Comment (draganmladjenovic) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: Shouldn't the constant be only 0? I fail to see how this optimization would be valid if we wish to produce 31, 63, or -1.

**Review by draganmladjenovic — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 2 above)
Comment (djtodoro) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: Not sure I follow.

Comment (djtodoro) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: Ah, yes.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 3 above)
Comment (djtodoro) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: Hmm, I agree. Let's avoid it for now. Looks like some tests were handwritten. I guess that the motivation behind was "whatever we chose (via `select`) in the case of ```0```, let's consider as a valid result". I think the motivation behind was taken from GCC (which does not imply to be correct).

**Review by draganmladjenovic — APPROVED** [review]

**Pull request merged**